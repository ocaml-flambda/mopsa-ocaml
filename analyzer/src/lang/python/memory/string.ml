(****************************************************************************)
(*                   Copyright (C) 2017 The MOPSA Project                   *)
(*                                                                          *)
(*   This program is free software: you can redistribute it and/or modify   *)
(*   it under the terms of the CeCILL license V2.1.                         *)
(*                                                                          *)
(****************************************************************************)

(** Methods of class str. *)

open Framework.Essentials
open Universal.Ast
open Ast
open Addr
open Addr_env

module Domain = struct
  type _ domain += D_python_memory_string : unit domain

  let id = D_python_memory_string
  let name = "python.memory.string"
  let identify : type a. a domain -> (unit, a) eq option = function
    | D_python_memory_string -> Some Eq
    | _ -> None

  let debug fmt = Debug.debug ~channel:name fmt

  let exec_interface = {export = []; import = []}
  let eval_interface = {export = [Zone.Z_py_value, any_zone]; import = []}

  let init _ _ flow = Some flow

  let mk_py_string s range =
    let mk_py_string_expr e range =
      let addr = {addr_kind = A_py_instance (find_builtin "str", None);
                  addr_uid = Universal.Heap.Pool.get_fresh ()} in
      debug "addr is %a (uid is %d)@\n" pp_addr addr addr.addr_uid;
      mk_py_object (addr, e) range
    in
    mk_py_string_expr (Universal.Ast.mk_string s range) range

  let eval zs exp man flow =
    let range = exp.erange in
    match ekind exp with
    (* 𝔼⟦ str.__new__(cls, arg) ⟧ *)
    | E_py_call({ekind = E_py_object ({addr_kind = A_py_function (F_builtin "str.__new__")}, _)}, cls :: args, []) ->
      Framework.Exceptions.panic_at range "str.__new__ not implemented"

    (* 𝔼⟦ s | s ∈ String ⟧ *)
    | E_constant (C_string s) ->
       Eval.singleton (mk_py_string s range) flow
       |> OptionExt.return

    (* 𝔼⟦ str.__len__(self) ⟧ *)
    | E_py_call({ekind = E_py_object ({addr_kind = A_py_function (F_builtin "str.__len__")}, _)}, [self], []) ->
      Framework.Exceptions.panic_at range "str.__len__ not implemented"

    (* 𝔼⟦ str.__eq__(self, other) ⟧ *)
    | E_py_call({ekind = E_py_object ({addr_kind = A_py_function (F_builtin "str.__eq__")}, _)}, [self; other], []) ->
      Framework.Exceptions.panic_at range "str.__eq__ not implemented"

    (* 𝔼⟦ str.__ne__(self, other) ⟧ *)
    | E_py_call({ekind = E_py_object ({addr_kind = A_py_function (F_builtin "str.__ne__")}, _)}, [self; other], []) ->
      Framework.Exceptions.panic_at range "str.__eq__ not implemented"

    | _ -> None


  let exec _ _ _ _ = None

  let ask _ _ _ = None

end

let () =
  Framework.Domains.Stateless.register_domain (module Domain)
