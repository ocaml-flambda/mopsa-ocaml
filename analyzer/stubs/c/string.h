#ifndef _STRING_H
#define _STRING_H

void *memset(void *s, int c, size_t n);

#endif
