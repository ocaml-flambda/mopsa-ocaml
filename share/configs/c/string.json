{
    "language": "c",
    "domain": {
	"seq": [
            // Universal iterators
            "universal.iterators.intraproc",
            "universal.iterators.loops",
            "universal.iterators.interproc.inlining",
            "universal.iterators.unittest",

	    // Stubs iterator
	    "stubs.iterator",

	    // C libraries
            "c.libs.libc",
            "c.libs.libmopsa",

	    // C iterators
            "c.iterators.program",
            "c.iterators.interproc",
            "c.iterators.goto",
            "c.iterators.switch",

            // C desugar
            "c.desugar.low_level",
            "c.desugar.loops",
            "c.desugar.record_copy",
            "c.desugar.stub_init",
            "c.desugar.cond_expr",

	    // Resouce management
            "universal.heap.recency.by_kind_callstack_range",
	    "c.resources.file",
	    "c.resources.common",

            // Memory model
	    {
                "apply": "c.memory.strings.length",
                "on": {
                    "nonrel": {
			"disjoint": [
			    "universal.numeric.values.intervals.float",
                            "universal.numeric.values.intervals.integer"
			]
		    }
                }
	    }
	]
    }
}
