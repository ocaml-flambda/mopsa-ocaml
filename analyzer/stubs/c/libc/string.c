/****************************************************************************/
/*                   Copyright (C) 2018 The MOPSA Project                   */
/*                                                                          */
/*   This program is free software: you can redistribute it and/or modify   */
/*   it under the terms of the CeCILL license V2.1.                         */
/*                                                                          */
/****************************************************************************/

#define __USE_GNU

/*
  libc stub
  based on header from glibc-2.27-r6
*/
#include <string.h>
#include "mopsa_libc_utils.h"

/*$$
 * predicate no_overlap(s1, s2, n):
 *   base(s1) != base(s2) or
 *   not ((unsigned char*)s1 <= (unsigned char*)s2 + n) or
 *   not ((unsigned char*)s2 <= (unsigned char*)s1 + n);
 */

/*$
 * requires: no_overlap(__src, __dest, __n);
 * requires: size(__src) >= __n;
 * requires: size(__dest) >= __n;
 * assigns: ((unsigned char*)__dest)[0, __n - 1];
 * ensures: forall int i in [0, __n - 1]: ((unsigned char*)__dest[i])' == (unsigned char*)__src[i];
 */
void *memcpy (void *__restrict __dest, const void *__restrict __src,
              size_t __n);

/*$
 * requires: size(__src) >= __n;
 * requires: size(__dest) >= __n;
 * assigns: ((unsigned char*)__dest)[0, __n - 1];
 * ensures: forall int i in [0, __n - 1]: ((unsigned char*)__dest[i])' == (unsigned char*)__src[i];
 */
void *memmove (void *__dest, const void *__src, size_t __n);


#if defined __USE_MISC || defined __USE_XOPEN

/*$
 * requires: no_overlap(__src, __dest, __n);
 * requires: size(__src) >= __n;
 * requires: size(__dest) >= __n;
 *
 * case "notfound" {
 *   assumes: forall int i in [0, __n - 1]: (unsigned char*)__src[i] != __c;
 *   assigns: ((unsigned char*)__dest)[0, __n - 1];
 *   ensures: forall int i in [0, __n - 1]: ((unsigned char*)__dest[i])' == (unsigned char*)__src[i];
 *   ensures: return == NULL;
 * }
 *
 * case "found" {
 *   assumes: exists int i in [0, __n - 1]: (unsigned char*)__src[i] == __c;
 *   assigns: ((unsigned char*)__dest)[0, __n - 1];
 *   ensures: exists int i in [0, __n - 1]: (
 *              (unsigned char*)__src[i] == __c and
 *              ((unsigned char*)__dest[i])' == __c and
 *              return == (unsigned char*)__src + i + 1 and
 *              forall int j in [0, i - 1]: (
 *                (unsigned char*)__src[j] != __c and
 *                ((unsigned char*)__dest[j])' == (unsigned char*)__src[j]
 *              ) and
 *              forall int j in [i, __n - 1]: ((unsigned char*)__dest[j])' == (unsigned char*)__dest[j]
 *            );
 * }
 */
void *memccpy (void *__restrict __dest, const void *__restrict __src,
		      int __c, size_t __n);

#endif /* Misc || X/Open.  */

/*$
 * requires: size(__s) >= __n;
 * assigns: ((unsigned char*)__s)[0, __n - 1];
 * ensures: forall int i in [0, __n - 1]: ((unsigned char*)__s[i])' == __c;
 * ensures: return == __s;
 */
void *memset (void *__s, int __c, size_t __n);

/*$
 * requires: size(__s1) >= __n;
 * requires: size(__s2) >= __n;
 *
 * case "equal" {
 *   assumes: forall int i in [0, __n - 1]: (unsigned char*)__s1[i] == (unsigned char*)__s2[i];
 *   ensures: return == 0;
 * }
 *
 * case "notequal" {
 *   assumes: exists int i in [0, __n - 1]: (unsigned char*)__s1[i] != (unsigned char*)__s2[i];
 *   ensures: exists int i in [0, __n - 1]: (
 *             (unsigned char*)__s1[i] != (unsigned char*)__s2[i] and
 *             forall int j in [0, i - 1]: (unsigned char*)__s1[j] == (unsigned char*)__s2[j] and
 *             ((unsigned char*)__s1[i] - (unsigned char*)__s2[i] > 0 implies return > 0) and
 *             ((unsigned char*)__s1[i] - (unsigned char*)__s2[i] < 0 implies return < 0)
 *            );
 * }
 */
int memcmp (const void *__s1, const void *__s2, size_t __n);

/*$
 * requires: size(__s) >= __n;
 *
 * case "found" {
 *   assumes: exists int i in [0, __n - 1]: (unsigned char*)__s[i] == __c;
 *   ensures: exists int i in [0, __n - 1]: (
 *              (unsigned char*)__s[i] == __c and
 *              forall int j in [0, i - 1]: (unsigned char*)__s[i] != __c and
 *              return == (unsigned char*)__s + i
 *            );
 * }
 *
 * case "notfound" {
 *   assumes: forall int i in [0, __n - 1]: (unsigned char*)__s[i] != __c;
 *   ensures: return == NULL;
 * }
 */
void *memchr (const void *__s, int __c, size_t __n);

#ifdef __USE_GNU

/*$
 * requires: size(__s) >= __n;
 *
 * case "found" {
 *   assumes: exists int i in [0, __n - 1]: (unsigned char*)__s[i] == __c;
 *   ensures: exists int i in [0, __n - 1]: (
 *              (unsigned char*)__s[i] == __c and
 *              forall int j in [0, i - 1]: (unsigned char*)__s[i] != __c and
 *              return == (unsigned char*)__s + i
 *            );
 * }
 *
 * case "notfound" {
 *   assumes: forall int i in [0, __n - 1]: (unsigned char*)__s[i] != __c;
 *   ensures: return == NULL;
 * }
 */
void *memrchr (const void *__s, int __c, size_t __n);

#endif

/*$
 * local: size_t src_len = strlen(__src);
 * requires: no_overlap(__src, __dest, src_len);
 * requires: size(__dest) >= src_len;
 * assigns: __dest[0, src_len - 1];
 * ensures: forall int i in [0, src_len - 1]: (__dest[i])' == __src[i];
 * ensures: return == __dest;
 */
char *strcpy (char *__restrict __dest, const char *__restrict __src);

/*$
 * local: size_t src_len = strlen(__src);
 * requires: no_overlap(__src, __dest, src_len);
 * requires: size(__dest) >= src_len or 
 *           size(__dest) >= __n;
 *
 * case "zero" {
 *   assumes: __n > src_len + 1;
 *   assigns: __dest[0, src_len];
 *   ensures: forall int i in [0, src_len]: (__dest[i])' == __src[i];
 *   ensures: return == __dest;
 * }
 * 
 * case "nozero" {
 *   assumes: __n <= src_len + 1;
 *   assigns: __dest[0, __n];
 *   ensures: forall int i in [0, __n - 1]: (__dest[i])' == __src[i];
 *   ensures: return == __dest;
 * }
 */
char *strncpy (char *__restrict __dest,
               const char *__restrict __src, size_t __n);

/*$
 * local: size_t src_len = strlen(__src);
 * local: size_t dest_len = strlen(__dest);
 * requires: no_overlap(__src, __dest, src_len);
 * requires: size(__dest) >= src_len + dest_len + 1;
 * assigns: __dest[0, src_len + dest_len + 1];
 * ensures: forall int i in [0, dest_len - 1]: (__dest[i])' == __dest[i];
 * ensures: forall int i in [0, src_len]: (__dest[i + dest_len])' == __src[i];
 * ensures: return == __dest;
 */
char *strcat (char *__restrict __dest, const char *__restrict __src);

/*$
 * local: size_t src_len = strlen(__src);
 * local: size_t dest_len = strlen(__dest);
 * requires: no_overlap(__src, __dest, src_len);
 * requires: size(__dest) >= src_len + dest_len + 1 or
 *           size(__dest) >= __n + dest_len + 1;
 * 
 * case "zero-found" {
 *   assumes: __n > src_len + 1;
 *   assigns: __dest[0, src_len + dest_len + 1];
 *   ensures: forall int i in [0, dest_len - 1]: (__dest[i])' == __dest[i];
 *   ensures: forall int i in [0, src_len]: (__dest[i + dest_len])' == __src[i];
 *   ensures: return == __dest;
 * }
 *
 * case "zero-not-found" {
 *   assumes: __n <= src_len + 1;
 *   assigns: __dest[0, __n + dest_len + 1];
 *   ensures: forall int i in [0, dest_len - 1]: (__dest[i])' == __dest[i];
 *   ensures: forall int i in [0, __n - 1]: (__dest[i + dest_len])' == __src[i];
 *   ensures: (__dest[dest_len + __n])' == 0;
 *   ensures: return == __dest;
 * }
 */
char *strncat (char *__restrict __dest, const char *__restrict __src,
               size_t __n);

/*$
 * local: size_t len1 = strlen(__s1);
 * local: size_t len2 = strlen(__s2);
 *
 * case "equal" {
 *   assumes: len1 == len2 and
 *            forall int i in [0, len1 - 1]: __s1[i] == __s2[i];
 *   ensures: return == 0;
 * }
 *
 * case "not-equal" {
 *   assumes: exists int i in [0, len1 - 1]: (
 *              i <= len2 - 1 and 
 *              __s1[i] != __s2[i]
 *            );
 *   ensures: exists int i in [0, len1 - 1]: (
 *              i <= len2 - 1 and 
 *              __s1[i] != __s2[i] and
 *              forall int j in [0, i - 1]: __s1[j] == __s2[j] and
 *             ((unsigned char*)__s1[i] - (unsigned char*)__s2[i] > 0 implies return > 0) and
 *             ((unsigned char*)__s1[i] - (unsigned char*)__s2[i] < 0 implies return < 0)
 *           );
 *  }
 */
int strcmp (const char *__s1, const char *__s2);

/*$
 * requires: size(__s1) >= __n or valid_string(__s1);
 * requires: size(__s2) >= __n or valid_string(__s2);
 *
 * case "equal" {
 *   assumes: exists int l in [0, __n - 1]: (
 *              (l == __n - 1 or __s1[l] == 0) and
 *              forall int i in [0, l]: __s1[i] == __s2[i]
 *   );
 *   ensures: return == 0;
 * }
 *
 * case "notequal" {
 *   assumes: exists int l in [0, __n - 1]: (
 *              __s1[l] != __s2[l] and
 *              forall int i in [0, l - 1]: (__s1[i] != 0 or __s2[i] != 0)
 *            );
 *   ensures: exists int l in [0, __n - 1]: (
 *              __s1[l] != __s2[l] and
 *              forall int i in [0, l]: __s1[i] == __s2[i] and
 *              ((unsigned char*)__s1[l] - (unsigned char*)__s2[l] > 0 implies return > 0) and
 *              ((unsigned char*)__s1[l] - (unsigned char*)__s2[l] < 0 implies return < 0)
 *            );
 *   }
 */
int strncmp (const char *__s1, const char *__s2, size_t __n);

/*$
 * requires: valid_string(__s1);
 * requires: valid_string(__s2);
 * assigns:  _errno;
 */
int strcoll (const char *__s1, const char *__s2);

/*$
 * requires: valid_string(__src);
 * assigns: __dest[0, __n - 1];
 * assigns: _errno;
 * ensures: return < __n implies (__dest[return])' == 0;
 */
size_t strxfrm (char *__restrict __dest,
                const char *__restrict __src, size_t __n);

#ifdef __USE_XOPEN2K8


// type __locale_t has been renamed local_t starting from glibv 2.26
#if __GLIBC_MINOR__ < 26

/*$
 * requires: valid_string(__s1);
 * requires: valid_string(__s2);
 * assigns:  _errno;
 */
int strcoll_l (const char *__s1, const char *__s2, __locale_t __l);

/*$
 * requires: valid_string(__src);
 * assigns: __dest[0, __n - 1];
 * assigns: _errno;
 * ensures: return < __n implies (__dest[return])' == 0;
 */
size_t strxfrm_l (char *__dest, const char *__src, size_t __n, __locale_t __l);

#else

/*$
 * requires: valid_string(__s1);
 * requires: valid_string(__s2);
 * assigns:  _errno;
 */
int strcoll_l (const char *__s1, const char *__s2, locale_t __l);

/*$
 * requires: valid_string(__src);
 * assigns: __dest[0, __n - 1];
 * assigns: _errno;
 * ensures: return < __n implies (__dest[return])' == 0;
 */
size_t strxfrm_l (char *__dest, const char *__src, size_t __n, locale_t __l);

#endif

#endif

#if defined __USE_XOPEN_EXTENDED || defined __USE_XOPEN2K8

/*$
 * local: size_t len = strlen(__s);
 *
 * case "success" {
 *   local: char* r = new Memory;
 *   ensures: return == r;
 *   ensures: size(return) == len + 1;
 *   ensures: forall int l in [0, len]: r[l] == __s[l];
 * }
 *
 * case "failure" {
 *   assigns: _errno;
 *   ensures: return == NULL;
 * }
 */
char *strdup (const char *__s);

#endif

#if defined __USE_XOPEN2K8

/*$
 * requires: size(__s) >= __n or valid_string(__s);
 *
 * case "success" {
 *   local: char* r = new Memory;
 *   ensures: return == r;
 *   ensures: exists int l in [0, __n]: (
 *              (l == __n or __s[l] == 0) and
 *              r[l] == 0 and
 *              size(return) == l + 1 and
 *              forall int i in [0, l - 1]: (
 *                __s[i] != 0 and
 *                r[i] == __s[i]
 *              )
 *            );
 * }
 *
 * case "failure" {
 *   assigns: _errno;
 *   ensures: return == NULL;
 * }
 */
char *strndup (const char *__s, size_t __n);

#endif

/*$
 * local: size_t len = strlen(__s);
 *
 * case "found" {
 *   assumes: exists int i in [0, len]: __s[i] == __c;
 *   ensures: exists int i in [0, len]: (
 *              __s[i] == __c and
 *              return == __s + i and
 *              forall int j in [0, i - 1]: __s[j] != __c
 *            );
 * }
 *
 * case "notfound" {
 *   assumes: forall int j in [0, len]: __s[j] != __c;
 *   ensures: return == NULL;
 * }
 */
char *strchr (const char *__s, int __c);

/*$
 * local: size_t len = strlen(__s);
 *
 * case "found" {
 *   assumes: exists int i in [0, len]: __s[i] == __c;
 *   ensures: exists int i in [0, len]: (
 *              __s[i] == __c and
 *              return == __s + i and
 *              forall int j in [i, len]: __s[j] != __c
 *            );
 * }
 *
 * case "notfound" {
 *   assumes: forall int j in [0, len]: __s[j] != __c;
 *   ensures: return == NULL;
 * }
 */
char *strrchr (const char *__s, int __c);


#ifdef __USE_GNU

char *strchrnul (const char *__s, int __c);

#endif

size_t strcspn (const char *__s, const char *__reject);

size_t strspn (const char *__s, const char *__accept);

char *strpbrk (const char *__s, const char *__accept);

char *strstr (const char *__haystack, const char *__needle);

char *strtok (char *__restrict __s, const char *__restrict __delim);

char *__strtok_r (char *__restrict __s,
                  const char *__restrict __delim,
                  char **__restrict __save_ptr);

#ifdef __USE_POSIX

char *strtok_r (char *__restrict __s, const char *__restrict __delim,
                char **__restrict __save_ptr);

#endif

#ifdef __USE_GNU

char *strcasestr (const char *__haystack, const char *__needle);

#endif

#ifdef __USE_GNU

void *memmem (const void *__haystack, size_t __haystacklen,
              const void *__needle, size_t __needlelen);

void *__mempcpy (void *__restrict __dest,
                 const void *__restrict __src, size_t __n);

void *mempcpy (void *__restrict __dest,
               const void *__restrict __src, size_t __n);

#endif

size_t strlen (const char *__s);

#ifdef	__USE_XOPEN2K8

size_t strnlen (const char *__string, size_t __maxlen);

#endif

char *strerror (int __errnum);

#if defined __USE_XOPEN2K && !defined __USE_GNU

int strerror_r (int __errnum, char *__buf, size_t __buflen);

#else

char *strerror_r (int __errnum, char *__buf, size_t __buflen);

#endif

#ifdef __USE_XOPEN2K8

char *strerror_l (int __errnum, __locale_t __l);

#endif

#ifdef __USE_MISC

void explicit_bzero (void *__s, size_t __n);

char *strsep (char **__restrict __stringp,
              const char *__restrict __delim);

#endif

#ifdef	__USE_XOPEN2K8

char *strsignal (int __sig);

char *__stpcpy (char *__restrict __dest, const char *__restrict __src);

char *stpcpy (char *__restrict __dest, const char *__restrict __src);

char *__stpncpy (char *__restrict __dest,
                 const char *__restrict __src, size_t __n);

char *stpncpy (char *__restrict __dest,
               const char *__restrict __src, size_t __n);

#endif

#ifdef	__USE_GNU

int strverscmp (const char *__s1, const char *__s2);

char *strfry (char *__string);

void *memfrob (void *__s, size_t __n);

# ifndef basename

char *basename (const char *__filename);

#endif

#endif
