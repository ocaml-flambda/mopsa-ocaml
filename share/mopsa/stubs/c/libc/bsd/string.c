/****************************************************************************/
/*                                                                          */
/* This file is part of MOPSA, a Modular Open Platform for Static Analysis. */
/*                                                                          */
/* Copyright (C) 2017-2019 The MOPSA Project.                               */
/*                                                                          */
/* This program is free software: you can redistribute it and/or modify     */
/* it under the terms of the GNU Lesser General Public License as published */
/* by the Free Software Foundation, either version 3 of the License, or     */
/* (at your option) any later version.                                      */
/*                                                                          */
/* This program is distributed in the hope that it will be useful,          */
/* but WITHOUT ANY WARRANTY; without even the implied warranty of           */
/* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the            */
/* GNU Lesser General Public License for more details.                      */
/*                                                                          */
/* You should have received a copy of the GNU Lesser General Public License */
/* along with this program.  If not, see <http://www.gnu.org/licenses/>.    */
/*                                                                          */
/****************************************************************************/

/*
  libc stub
  utility functions from BSD system (libbsd)
*/

#include <bsd/string.h>
#include "mopsa_libc_utils.h"

/*$
 * requires: valid_bytes(__dst, __size);
 * requires: __size > 0;
 * local: size_t src_len = strlen(__src);
 * ensures: return == src_len;
 *
 * case "small" {
 *   assumes: __size > src_len;
 *   assigns: __dst[0, src_len];
 *   ensures: forall size_t i in [0, src_len]: (__dst[i])' == __src[i];
 * }
 *
 * case "large" {
 *   assumes: __size <= src_len;
 *   assigns: __dst[0, __size);
 *   ensures: forall size_t i in [0, __size-1): (__dst[i])' == __src[i];
 *   ensures: (__dst[__size-1])' == 0;
 * }
 */
size_t strlcpy(char *__dst, const char *__src, size_t __size);

/*$
 * requires: valid_bytes(__dst, __size);
 * requires: __size > 0;
 * local: size_t src_len = strlen(__src);
 * local: size_t dst_len = strlen(__dst);
 * ensures: return == src_len + dst_len;
 *
 * case "small" {
 *   assumes: __size > src_len + dst_len;
 *   assigns: __dst[0, __size);
 *   ensures: forall size_t i in [0, dst_len): (__dst[i])' == __dst[i];
 *   ensures: forall size_t i in [0, src_len]: (__dst[dst_len + i])' == __src[i];
 * }
 *
 * case "large" {
 *   assumes: __size > dst_len and __size <= src_len + dst_len;
 *   assigns: __dst[0, __size);
 *   ensures: forall size_t i in [0, dst_len): (__dst[i])' == __dst[i];
 *   ensures: forall size_t i in [0, __size-dst_len-1): (__dst[dst_len + i])' == __src[i];
 *   ensures: (__dst[__size-1])' == 0;
 * }
 *
 * case "too_large" {
 *   assumes: __size <= dst_len;
 * }
 */
size_t strlcat(char *__dst, const char *__src, size_t __size);

/*$
 * local: size_t len1 = strlen(__str);
 * local: size_t len2 = strlen(__find);
 *
 * case "small" {
 *   assumes: len1 < __str_len;
 *   ensures: return == NULL or
 *            (len1 >= len2 and in_bytes(return, __str, len1 - len2));
 * }
 *
 * case "large" {
 *   assumes: len1 >= __str_len;
 *   ensures: return == NULL or
 *            (__str_len >= len2 and in_bytes(return, __str, __str_len - len2));
 * }
 */
char *strnstr(const char *__str, const char *__find, size_t __str_len);
//
/*$
 * requires: valid_bytes(__str, 12);
 * assigns: __str[0, 11];
 * ensures: (__str[11])' == 0;
 */
void strmode(mode_t mode, char *__str);
