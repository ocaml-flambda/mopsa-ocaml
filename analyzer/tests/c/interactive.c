#include <string.h>

int f(int x) {
  char *s = "";
  size_t b = strlen(s);
  return b;
}

int main() {
  char *s = NULL;
  int x = 42;
  f(3);
}
