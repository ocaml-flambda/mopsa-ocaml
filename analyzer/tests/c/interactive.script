b f
c
b 6
c
p x, s
p x:main
p x:f
p x:*,s:*
p s:main
p s:f
mopsa_bt
f
