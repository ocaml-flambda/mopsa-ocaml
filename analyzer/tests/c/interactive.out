    3  int f(int x) {
    4    char *s = "";
    5    size_t b = strlen(s);
 [1;38;5;118m‚ñ∫[0m [1;38;5;118m 6[0m  [1;38;5;118m  return b;[0m
    7  }
    8  
    9  int main() {
   10    char *s = NULL;
   11    int x = 42;
[1;38;5;45mùïä[0m [1;38;5;45m‚ü¶[0m return b; [1;38;5;45m‚üß[0m
 float-itv ‚à™ int-itv : 
    offset‚¶És‚¶Ñ : [0,0],
    x : [3,3] ,
 pointers : 
    s : {  ""  }  
    s:interactive.c:10.2-16 : { NULL },
    s:interactive.c:4.2-14 : { "" }  
 float-itv ‚à™ int-itv : 
    x : [42,42]  
 float-itv ‚à™ int-itv : 
    x : [3,3]  
 float-itv ‚à™ int-itv : 
    offset‚¶És‚¶Ñ : [0,0],
    x : [3,3],
    x : [42,42] ,
 pointers : 
    s : {  ""  },
    s : {  NULL  }  
 pointers : 
    s : {  NULL  }  
 float-itv ‚à™ int-itv : 
    offset‚¶És‚¶Ñ : [0,0] ,
 pointers : 
    s : {  ""  }  
Raised by primitive operation at Interactive__Terminal.Make.pp_output_command in file "analyzer/framework/engines/interactive/terminal.ml", line 618, characters 47-83
Called from Interactive__Terminal.Make.read_command in file "analyzer/framework/engines/interactive/terminal.ml", line 855, characters 6-54
Called from Interactive__Engine.Make.interact in file "analyzer/framework/engines/interactive/engine.ml", line 228, characters 14-78
Called from Interactive__Engine.Make.interact_or_apply_action in file "analyzer/framework/engines/interactive/engine.ml", line 286, characters 10-30
Called from Core__Post.bind.(fun) in file "analyzer/framework/core/post.ml", line 53, characters 37-45
Called from Core__Post.bind in file "analyzer/framework/core/post.ml", line 53, characters 2-46
Called from Stdlib__List.fold_left in file "list.ml", line 121, characters 24-34
Called from Universal_iterators__Intraproc.Domain.exec in file "analyzer/languages/universal/iterators/intraproc.ml", line 163, characters 19-98
Called from Combiners_common__Common.cascade_stateless_call.(fun) in file "analyzer/framework/combiners/common/common.ml", line 230, characters 13-29
Called from Combiners_common__Common.cascade_stateless_call.(fun) in file "analyzer/framework/combiners/common/common.ml", line 230, characters 13-29
Called from Combiners_common__Common.cascade_call.(fun) in file "analyzer/framework/combiners/common/common.ml", line 187, characters 13-44
Called from Combiners_common__Common.cascade_call.(fun) in file "analyzer/framework/combiners/common/common.ml", line 187, characters 13-44
Called from Core__Cache.Make.exec in file "analyzer/framework/core/cache.ml", line 131, characters 19-34
Called from Toplevel.Make.exec.(fun) in file "analyzer/framework/toplevel/toplevel.ml", line 235, characters 16-52
